Split up .NET fundamentals TOC

gewarren opened the pull request:
Contributes to #33349 

This doesn't contain changes to the L2 header since that [happens in a different repo](https://github.com/MicrosoftDocs/DocsRoot/pull/1464/files).

Previews:

- [Fundamentals](https://review.learn.microsoft.com/en-us/dotnet/fundamentals/?branch=pr-en-us-33610)
- [Advanced programming](https://review.learn.microsoft.com/en-us/dotnet/standard/asynchronous-programming-patterns/?branch=pr-en-us-33610)
- [Tools and diagnostics](https://review.learn.microsoft.com/en-us/dotnet/toc/tools-diagnostics/?branch=pr-en-us-33610)
- [DevOps and testing](https://review.learn.microsoft.com/en-us/dotnet/devops/github-actions-overview?branch=pr-en-us-33610)
- [Migration guide](https://review.learn.microsoft.com/en-us/dotnet/core/porting/?branch=pr-en-us-33610)
- [Data access](https://review.learn.microsoft.com/en-us/dotnet/standard/data/xml/?branch=pr-en-us-33610)
- [Security](https://review.learn.microsoft.com/en-us/dotnet/standard/security/?branch=pr-en-us-33610)

gewarren committed:
split .net fundamentals toc

gewarren committed:
add to landing pages

gewarren committed:
fix bad links

gewarren committed:
rearrange fundamentals landing page

gewarren committed:
update breadcrumbs

gewarren committed:
fix bad links

gewarren committed:
reorder tools-diagnostics landing page

tdykstra reviewed (commented):
If you're in one of these TOCs how do you see what others are available, and how do you get to one of the others? If, for example, you're on the Fundamentals TOC, how do you see that there's a separate Tools and diagnostics TOC and how do you get to it? Or the reverse - how do you go from Tools and diagnostics to Fundamentals?

gewarren commented:
> If you're in one of these TOCs how do you see what others are available, and how do you get to one of the others? If, for example, you're on the Fundamentals TOC, how do you see that there's a separate Tools and diagnostics TOC and how do you get to it? Or the reverse - how do you go from Tools and diagnostics to Fundamentals?

This would be either from the new proposed L2 header menu ("Features") or by following the breadcrumb up the chain to ".NET" and the .NET hub page. Which will need to be updated to list each of the new proposed TOCs if we decide to move forward with this.

gewarren commented on docs/toc/performance-security/index.yml:
```diff
@@ -0,0 +1,24 @@
+### YamlMime:Landing
+
+title: .NET performance and security documentation
```
They aren't really related but are both concepts that are secondary to just getting your app to "work". I almost think we can get rid of "performance" as a heading altogether and move the memory management articles into the "Advanced programming" TOC. For security, we could either add that to "Fundamentals" or "Advanced programming".

BillWagner commented:
I like the strategy here a lot. The .NET ecosystem is quite large, and we need to have a solid organization for all the content to make it navigable for all readers.

This places a lot more emphasis on the L2 header for navigation, along with the breadcrumb header for knowledge of place. That's a good way forward, but we may need to consider how we raise more awareness of those elements.

On the specific breakdown:
- I wonder if performance should move to "Advanced programming techniques". 
- I'd like to have more visibility in the data section for other data storage models. In particular, we should link to the various .NET toolkits for Azure storage (tables, CosmosDB, etc).
- Should the JSON serialization articles be under data access? Or should XML data access be under Serialization in the Fundamentals section? (They seem related, and now they are separated).

As we iron these out, this is a great direction!

tdykstra commented:
> This places a lot more emphasis on the L2 header for navigation, along with the breadcrumb header for knowledge of place. That's a good way forward, but we may need to consider how we raise more awareness of those elements.

Maybe at the end of each TOC have a node that says something like "More..." and links to a page that shows screenshots of the L2 and breadcrumbs being used. That way there wouldn't need to be any TOC context shifting.

> * I wonder if performance should move to "Advanced programming techniques".

+1

> * I'd like to have more visibility in the data section for other data storage models. In particular, we should link to the various .NET toolkits for Azure storage (tables, CosmosDB, etc).

+1

> * Should the JSON serialization articles be under data access? Or should XML data access be under Serialization in the Fundamentals section? (They seem related, and now they are separated).

XML data access seems to me to fit best in the Data access section, but serialization doesn't. I would keep Serialization in Fundamentals. We can cross-link between the XML serialization and XML data access articles.

gewarren committed:
flesh out landing pages and respond to feedback

gewarren commented:
Okay, I moved memory management articles to the advanced programming TOC and left security in its own TOC. I moved the memory/span types and SIMD types articles to fundamental coding components in the main TOC where we talk about other categories of types.

gewarren committed:
fix broken links

BillWagner reviewed (approved):
This really looks great @gewarren 

I'm happy with the changes you made. I'm ready to :shipit:.
